From ca087d5b0e8cd4ac596c161dbf569640c7754b46 Mon Sep 17 00:00:00 2001
From: Klemens Nanni <kl3@posteo.org>
Date: Sat, 7 Jan 2017 02:18:09 +0100
Subject: [PATCH] Update 'mediakeys' patch to 20170111

---
 slock.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/slock.c b/slock.c
index ac26cc8..4d81ca6 100644
--- a/slock.c
+++ b/slock.c
@@ -186,7 +186,7 @@ readpw(Display *dpy, struct xrandr *rr, struct lock **locks, int nscreens,
 			case XF86XK_AudioStop:
 			case XF86XK_AudioPrev:
 			case XF86XK_AudioNext:
-				XSendEvent(dpy, locks[0]->root, True, KeyPressMask, &ev);
+				XSendEvent(dpy, DefaultRootWindow(dpy), True, KeyPressMask, &ev);
 				break;
 			case XK_Return:
 				passwd[len] = '\0';
-- 
2.11.0

